clients/src/main/java/org/apache/kafka/common/record/MemoryRecords.java

```diff
@@ -168,38 +177,33 @@ private static FilterResult filterTo(TopicPartition partition, Iterable<MutableR
             // allow for the possibility that a previous version corrupted the log by writing a compressed record batch
             // with a magic value not matching the magic of the records (magic < 2). This will be fixed as we
             // recopy the messages to the destination buffer.
-
             byte batchMagic = batch.magic();
+            // we want to check if the delete horizon has been set or stayed the same
             boolean writeOriginalBatch = true;
+            boolean containsTombstonesOrMarker = false;
             List<Record> retainedRecords = new ArrayList<>();
 
-            try (final CloseableIterator<Record> iterator = batch.streamingIterator(decompressionBufferSupplier)) {
-                while (iterator.hasNext()) {
-                    Record record = iterator.next();
-                    filterResult.messagesRead += 1;
-
-                    if (filter.shouldRetainRecord(batch, record)) {
-                        // Check for log corruption due to KAFKA-4298. If we find it, make sure that we overwrite
-                        // the corrupted batch with correct data.
-                        if (!record.hasMagic(batchMagic))
-                            writeOriginalBatch = false;
-
-                        if (record.offset() > maxOffset)
-                            maxOffset = record.offset();
-
-                        retainedRecords.add(record);
-                    } else {
-                        writeOriginalBatch = false;
-                    }
-                }
-            }
+            final BatchIterationResult iterationResult = iterateOverBatch(batch, decompressionBufferSupplier, filterResult, filter,
+                                                                          batchMagic, writeOriginalBatch, maxOffset, retainedRecords);
+            containsTombstonesOrMarker = iterationResult.containsTombstonesOrMarker();
+            writeOriginalBatch = iterationResult.shouldWriteOriginalBatch();
+            maxOffset = iterationResult.maxOffset();
 
             if (!retainedRecords.isEmpty()) {
-                if (writeOriginalBatch) {
+                // we check if the delete horizon should be set to a new value
+                // in which case, we need to reset the base timestamp and overwrite the timestamp deltas
+                // if the batch does not contain tombstones, then we don't need to overwrite batch
+                boolean needToSetDeleteHorizon = batch.magic() >= 2 && containsTombstonesOrMarker && !batch.deleteHorizonSet();
```

Comment:
You were correct earlier that for a control Marker, deleteHorizon should only be set after transactional records before the marker have already been removed. So, we can't just set needToSetDeleteHorizon based on containsTombstonesOrMarker. Also, I still feel that retrieveDeleteHorizon() is a bit weird since it mixes deleteHorizonMs that's already set with the deleteHorizonMs to be set.

So, perhaps it's clearer if we instead have a method containEmptyMarker() that simply passes along the return value of shouldDiscardBatch().

Then `needToSetDeleteHorizon = batch.magic() >= 2 && (containEmptyMarker || containsTombstones) && !batch.deleteHorizonSet())`.

If we need to set deleteHorizon, deleteHorizonMs can be computed off tombstoneRetentionMs, which can be passed into filterTo().

Reply:
Alright, sounds cool. This actually makes sense. I got it done.